File: app/src/main/java/co/tpcreative/supersafe/ui/privates/PrivatePresenter.java
package co.tpcreative.supersafe.ui.privates;
import com.snatik.storage.Storage;
import java.util.ArrayList;
import java.util.List;
import co.tpcreative.supersafe.common.controller.ServiceManager;
import co.tpcreative.supersafe.common.entities.ItemEntity;
import co.tpcreative.supersafe.common.entities.MainCategoryEntity;
import co.tpcreative.supersafe.common.helper.SQLHelper;
import co.tpcreative.supersafe.common.presenter.BaseView;
import co.tpcreative.supersafe.common.presenter.Presenter;
import co.tpcreative.supersafe.common.services.SuperSafeApplication;
import co.tpcreative.supersafe.common.util.Utils;
import co.tpcreative.supersafe.model.EnumDelete;
import co.tpcreative.supersafe.model.EnumFormatType;
import co.tpcreative.supersafe.model.EnumStatus;
import co.tpcreative.supersafe.common.entities.InstanceGenerator;
import co.tpcreative.supersafe.model.ItemModel;
import co.tpcreative.supersafe.model.MainCategoryModel;

public class PrivatePresenter extends Presenter<BaseView> {
    protected List<MainCategoryModel> mList;
    protected Storage storage;
    private static final String TAG = PrivatePresenter.class.getSimpleName();
    public PrivatePresenter(){
        mList = new ArrayList<>();
    }

    public void  getData(){
        BaseView view = view();
        mList = SQLHelper.getList();
        storage = new Storage(SuperSafeApplication.getInstance());
        view.onSuccessful("Successful", EnumStatus.RELOAD);
    }

    public void onDeleteAlbum(int position){
        BaseView view = view();
        try {
            final MainCategoryModel main = mList.get(position);
            if (main!=null){
                final List<ItemModel> mListItems = SQLHelper.getListItems(main.categories_local_id,false);
                for (ItemModel index : mListItems){
                    index.isDeleteLocal = true;
                    SQLHelper.updatedItem(index);
                }
                main.isDelete = true;
                SQLHelper.updateCategory(main);
            }
        }
        catch (Exception e){
            e.printStackTrace();
        }
        finally {
            getData();
            ServiceManager.getInstance().onPreparingSyncData();
        }
    }

    public void onEmptyTrash(){
        BaseView view = view();
        try {
            final List<ItemModel> mList = SQLHelper.getDeleteLocalListItems(true, EnumDelete.NONE.ordinal(),false);
            for (int i = 0 ;i <mList.size();i++){
                EnumFormatType formatTypeFile = EnumFormatType.values()[mList.get(i).formatType];
                if (formatTypeFile == EnumFormatType.AUDIO && mList.get(i).global_original_id==null){
                    SQLHelper.deleteItem(mList.get(i));
                }
                else if (formatTypeFile == EnumFormatType.FILES && mList.get(i).global_original_id==null){
                    SQLHelper.deleteItem(mList.get(i));
                }
                else if (mList.get(i).global_original_id==null & mList.get(i).global_thumbnail_id == null){
                    SQLHelper.deleteItem(mList.get(i));
                }
                else{
                    mList.get(i).deleteAction = EnumDelete.DELETE_WAITING.ordinal();
                    SQLHelper.updatedItem(mList.get(i));
                    Utils.Log(TAG,"ServiceManager waiting for delete");
                }
                storage.deleteDirectory(SuperSafeApplication.getInstance().getSupersafePrivate()+mList.get(i).items_id);
            }
        }
        catch (Exception e){
            e.printStackTrace();
        }
        finally {
            getData();
            ServiceManager.getInstance().onPreparingSyncData();
        }
    }
}

File: app/src/main/java/co/tpcreative/supersafe/ui/privates/PrivatePresenter.kt
package co.tpcreative.supersafe.ui.privates
import android.app.Activity
import android.content.Context
import co.tpcreative.supersafe.R
import co.tpcreative.supersafe.common.controller.ServiceManager
import co.tpcreative.supersafe.common.controller.SingletonManagerProcessing
import co.tpcreative.supersafe.common.helper.SQLHelper
import co.tpcreative.supersafe.common.presenter.BaseView
import co.tpcreative.supersafe.common.presenter.Presenter
import co.tpcreative.supersafe.common.services.SuperSafeApplication
import co.tpcreative.supersafe.common.util.Utils
import co.tpcreative.supersafe.model.*
import com.google.android.gms.auth.GoogleAuthException
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.googleapis.extensions.android.gms.auth.UserRecoverableAuthIOException
import com.google.gson.Gson
import com.snatik.storage.Storage
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import java.io.IOException
import java.util.*
import java.util.concurrent.Callable

class PrivatePresenter : Presenter<BaseView<EmptyModel>>() {
    var mList: MutableList<MainCategoryModel>?
    protected var storage: Storage? = null
    fun getData() {
        val view: BaseView<EmptyModel>? = view()
        mList = SQLHelper.getList()
        storage = Storage(SuperSafeApplication.getInstance())
        view?.onSuccessful("Successful", EnumStatus.RELOAD)
    }

    fun onDeleteAlbum(position: Int) {
        try {
            val main: MainCategoryModel? = mList?.get(position)
            main?.let {
                val mListItems: MutableList<ItemModel>? = SQLHelper.getListItems(it.categories_local_id, false)
                mListItems?.let {item ->
                    for (index in item) {
                        index.isDeleteLocal = true
                        SQLHelper.updatedItem(index)
                    }
                    it.isDelete = true
                    SQLHelper.updateCategory(it)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            getData()
            ServiceManager.getInstance()?.onPreparingSyncData()
        }
    }

    fun onEmptyTrash() {
        subscriptions?.add(Observable.fromCallable(Callable {
            try {
                val mList: MutableList<ItemModel>? = SQLHelper.getDeleteLocalListItems(true, EnumDelete.NONE.ordinal, false)
                mList?.let {
                    for (index in it) {
                        val formatTypeFile = EnumFormatType.values()[index.formatType]
                        if (formatTypeFile == EnumFormatType.AUDIO && index.global_original_id == null) {
                            SQLHelper.deleteItem(index)
                        } else if (formatTypeFile == EnumFormatType.FILES && index.global_original_id == null) {
                            SQLHelper.deleteItem(index)
                        } else if ((index.global_original_id == null) and (index.global_thumbnail_id == null)) {
                            SQLHelper.deleteItem(index)
                        } else {
                            index.deleteAction = EnumDelete.DELETE_WAITING.ordinal
                            SQLHelper.updatedItem(index)
                            Utils.Log(TAG, "ServiceManager waiting for delete")
                        }
                        storage?.deleteDirectory(SuperSafeApplication.getInstance().getSuperSafePrivate() + index.items_id)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
             ""
        }).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribe { response: String? ->
                    getData()
                    ServiceManager.getInstance()?.onPreparingSyncData()
                })
        Utils.Log(TAG,"empty data")
    }

    companion object {
        private val TAG = PrivatePresenter::class.java.simpleName
    }

    init {
        mList = ArrayList<MainCategoryModel>()
    }
}